### PR TITLE
CWC-1507 + System test improvements

### PR DESCRIPTION
**Pull request by ymarcus93**
Related PRs: https://github.com/bastionzero/webshell-common-ts/pull/62, https://github.com/bastionzero/bzero/pull/32, https://github.com/bastionzero/webshell-backend/pull/1014, https://github.com/bastionzero/charts/pull/9

## Description of the change

CWC-1507:
- Pass target/agent's base64-encoded public key to the daemon

System test improvements:
- Update AmazonLinux2 and BzeroVTAL2TestImage DigitalOcean images. git
  was broken in the old one
- Add `SYSTEM_TEST_TAGS` environment variable. Optionally set this variable to
  comma-separated list of tags to override default tag of `"system-tests"`
- Add `BZERO_AGENT_BRANCH` environment variable. Optionally set this
  variable to compile the `bzero-agent` from a specific `github.com/bastionzero/bzero` remote
  branch when running the VT tests.
- Add ability to test specific kube agent image hosted @ DigitalOcean. Specify `BZERO_KUBE_AGENT_IMAGE` to do this.
- [Use dots-wrapper typescript library instead of digitalocean-js](https://github.com/bastionzero/zli/commit/06789f558646183d4da191410b75b3cd51f8e176)

## Relevant release note information

Release Notes: CWC-1507 + System test improvements

## Related JIRA tickets

Relates to JIRA: CWC-1507, CWC-1545

## Have you considered the security impacts?

Does this PR have any security impact?

- [x] Yes
- [ ] No

If yes, please explain:

See security impact @ https://github.com/bastionzero/bzero/pull/32 and @ https://github.com/bastionzero/webshell-backend/pull/1014

### CODE REVIEW

**Comment by ymarcus93** [comment]
System tests will fail until we release the new helm chart: https://github.com/bastionzero/charts/pull/9

I've tested the new chart by running system tests locally pointing to chart directory on my machine.

Tested:

**NEW DAEMON - OLD AGENT**
Kube - Passed
VT - Passed

**OLD DAEMON - NEW AGENT**
Kube - Passed
VT - Passed

**NEW DAEMON - NEW AGENT**
Kube - Passed
VT - Passed

File: src/system-tests/tests/system-test.ts [hunk 1]
```diff
@@ -277,39 +336,75 @@ async function createDOTestClusters() {
...
+
     let registerCommand: string;
     let initBlock: string = '';
     switch(testTarget.installType) {
     case 'pm':
         registerCommand = `${packageName} --serviceUrl ${configService.serviceUrl()} -registrationKey "${registrationApiKeySecret}" -envName "${envName}"`;
         break;
     case 'pm-vt':
-        registerCommand = `${packageName} --serviceUrl ${configService.serviceUrl()} -registrationKey "${registrationApiKeySecret}" -environmentName "${envName}"`;
+        const executableName = shouldBuildFromSource ? './root/bzero/bctl/agent/agent' : packageName;
```
Comment (lipmas) [inline src/system-tests/tests/system-test.ts:406]: shouldnt we also be using the custom executable name for the `pm` case as well? I think this should be moved outside of the switch statement

Comment (lipmas) [reply src/system-tests/tests/system-test.ts:406]: nevermind i see now we only want to allow build from  source for new agent

File: src/system-tests/tests/system-test.ts [hunk 2]
```diff
@@ -344,11 +439,17 @@ sudo systemctl restart postgresql-12
         return _exhaustiveCheck;
     }
 
+    let endBlock : string = ''
+    if (shouldBuildFromSource) {
+        endBlock = './root/bzero/bctl/agent/agent &'
+    }
```
Comment (lipmas) [inline src/system-tests/tests/system-test.ts:445]: maybe its more clear to just add this to the `install` block

File: src/system-tests/digital-ocean/digital-ocean-kube-service.ts [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
 import { KubeHttpService } from '../../http-services/targets/kube/kube.http-services';
 import { KubeClusterSummary } from '../../../webshell-common-ts/http/v2/target/kube/types/kube-cluster-summary.types';
 import { PolicyHttpService } from '../../../src/http-services/policy/policy.http-services';
 import { TargetStatus } from '../../../webshell-common-ts/http/v2/target/types/targetStatus.types';
+import { createApiClient } from 'dots-wrapper';
+import { ICreateKubernetesClusterNodePoolApiRequest, IKubernetesCluster } from 'dots-wrapper/dist/kubernetes';
 
 export class DigitalOceanKubeService {
-    private doClient: DigitalOcean;
+    private doClient;
```
[the remark this answers is not in the document]
Comment (ymarcus93) [reply src/system-tests/digital-ocean/digital-ocean-kube-service.ts:15]: Oh yeah I wish I could put an explicit type, but for some reason this library does not have a type alias for the client (or at least I couldn't find one).

Here is what the compiler infers it to be:

```
DigitalOceanKubeService.doClient: Readonly<{
    _options: Readonly<{
        endpoint: string;
        requestTimeoutInMs: number;
    }>;
    account: Readonly<{
        getAccount: () => Promise<Readonly<GetAccountResponse>>;
    }>;
    ... 22 more ...;
    vpc: Readonly<...>;
}>
```

Notice the `... 22 more`. So yeah, I don't think we want to put an explicit type here unless the library has some alias for this large type

[the remark this answers is not in the document]
Comment (ymarcus93) [reply src/system-tests/digital-ocean/digital-ocean-kube-service.ts:15]: https://github.com/pjpimentel/dots/blob/f2092d9f53b1f00abe87ac30e1a5002718b38342/src/common/create-api-client/create-api-client.ts#L9-L326

File: src/system-tests/tests/system-test.ts (hunk 2 above)
Comment (ymarcus93) [reply src/system-tests/tests/system-test.ts]: [33739c8](https://github.com/bastionzero/zli/pull/316/commits/33739c8bb5e0bd40b2675deb8562fbd628d199f5)

File: src/system-tests/tests/utils/helm/helm-utils.ts [hunk 4]
```diff
@@ -37,7 +42,13 @@ export interface MultiStringValue {
  * @param variables A dictionary of key value strings to pass as value overrides
```
Comment (sidpremkumar) [inline src/system-tests/tests/utils/helm/helm-utils.ts:42]: Update doctoring if you are updating the function?

File: src/system-tests/tests/system-test.ts [hunk 5]
```diff
@@ -63,6 +64,30 @@ const KUBE_ENABLED = process.env.KUBE_ENABLED ? (process.env.KUBE_ENABLED === 't
 const VT_ENABLED = process.env.VT_ENABLED ? (process.env.VT_ENABLED === 'true') : true;
 const SSM_ENABLED =  process.env.SSM_ENABLED ? (process.env.SSM_ENABLED === 'true') : true;
 
+const systemTestTags = process.env.SYSTEM_TEST_TAGS ? process.env.SYSTEM_TEST_TAGS.split(',').filter(t => t != '') : ['system-tests'];
+
+// Set this environment variable to compile agent from specific remote branch
+const bzeroAgentBranch = process.env.BZERO_AGENT_BRANCH;
+// Go version to use when compiling vt bzero agent
+const goVersion = 'go1.17.7.linux-amd64';
```
Comment (sidpremkumar) [inline src/system-tests/tests/system-test.ts:72]: maybe a ref here so we can update in the future?

**Review by sidpremkumar — APPROVED** [review]
nits, but LGTM

File: src/system-tests/tests/system-test.ts (hunk 5 above)
Comment (ymarcus93) [reply src/system-tests/tests/system-test.ts:73]: [eb1a28f](https://github.com/bastionzero/zli/pull/316/commits/eb1a28f5a51b03f26d69ad42ff5e0c9a67d445b8)

File: src/system-tests/tests/utils/helm/helm-utils.ts (hunk 4 above)
Comment (ymarcus93) [reply src/system-tests/tests/utils/helm/helm-utils.ts:43]: [eb1a28f](https://github.com/bastionzero/zli/pull/316/commits/eb1a28f5a51b03f26d69ad42ff5e0c9a67d445b8)

**Pull request merged**